src/functionsDelete.ts

```diff
@@ -0,0 +1,41 @@
+import * as helper from "./functionsDeployHelper";
+import { Queue } from "./throttler/queue";
+import * as tasks from "./deploy/functions/tasks";
+import { DeploymentTimer } from "./deploy/functions/deploymentTimer";
+import { ErrorHandler } from "./deploy/functions/errorHandler";
+
+export async function deleteFunctions(
+  functionsNamesToDelete: string[],
+  scheduledFunctionNamesToDelete: string[],
+  projectId: string,
+  appEngineLocation: string
+): Promise<void> {
+  const cloudFunctionsQueue = new Queue<() => Promise<void>, void>({});
+  const schedulerQueue = new Queue<() => Promise<void>, void>({});
+  const timer = new DeploymentTimer();
+  const errorHandler = new ErrorHandler();
+  const taskParams = {
+    projectId,
+    timer,
+    errorHandler,
+  };
+  functionsNamesToDelete.forEach((fnName) => {
+    const deleteFunctionTask = tasks.deleteFunctionTask(taskParams, fnName);
+    cloudFunctionsQueue.run(deleteFunctionTask);
+  });
+  scheduledFunctionNamesToDelete.forEach((fnName) => {
+    const deleteSchedulerTask = tasks.deleteScheduleTask(taskParams, fnName, appEngineLocation);
+    schedulerQueue.run(deleteSchedulerTask);
+  });
+  const queuePromises = [cloudFunctionsQueue.wait(), schedulerQueue.wait()];
+
+  cloudFunctionsQueue.close();
+  schedulerQueue.close();
+  cloudFunctionsQueue.process();
+  schedulerQueue.process();
+
+  await Promise.all(queuePromises);
```

Comment:
Isn't there a bug here where we could fail to delete a schedule but succeed in deleting a function? The next time the user tried to do a delete operation we wouldn't see the function so we wouldn't see its schedule and we wouldn't ever clean up the customer's project.